src/ui/view/dataverse/component/generalMetadata/dateInterval.tsx

```diff
@@ -0,0 +1,83 @@
+import { useTranslation } from 'react-i18next'
+import * as E from 'fp-ts/Either'
+import { pipe } from 'fp-ts/function'
+import {
+  convertToLocalizedDateIfISODateTime,
+  validateISODateRange
+} from '@/util/isoDateTime/isoDateTime'
+import { activeLanguageWithDefault } from '@/ui/languages/languages'
+
+const FormattedDateSpan = ({
+  localizedDate,
+  label
+}: {
+  label: 'from' | 'to'
+  localizedDate: string
+}): JSX.Element => {
+  const { t } = useTranslation('common')
+  return (
+    <span>
+      {t(label)}&nbsp;
+      {localizedDate}
+    </span>
+  )
+}
+const InvalidDateFallback = (): JSX.Element => {
+  const { t } = useTranslation('generalMetadata')
+  return <span>{t('generalMetadata.invalidDate')}</span>
+}
+
+const handleLocalizedEndDate = (localizedEndDate: E.Either<Error, string>): JSX.Element =>
+  pipe(
+    localizedEndDate,
+    E.fold<Error, string, JSX.Element>(
+      () => <InvalidDateFallback />,
+      localizedEndDate => <FormattedDateSpan label="to" localizedDate={localizedEndDate} />
+    )
+  )
+
+const handleDateRange = (
+  startDate: string,
+  endDate: string,
+  localizedStartDate: string,
+  localizedEndDate: string
+): JSX.Element =>
+  pipe(
+    validateISODateRange(startDate, endDate),
+    E.fold(
+      () => <InvalidDateFallback />,
+      () => (
+        <>
+          <FormattedDateSpan label="from" localizedDate={localizedStartDate} />
+          <FormattedDateSpan label="to" localizedDate={localizedEndDate} />
+        </>
+      )
+    )
+  )
+
+type DateIntervalProps = {
+  startDate: string
+  endDate: string
+}
+
+export const DateInterval = ({ startDate, endDate }: DateIntervalProps): JSX.Element => {
+  const lng = activeLanguageWithDefault().lng
+  const localizedStartDate = convertToLocalizedDateIfISODateTime(startDate, lng)
+  const localizedEndDate = convertToLocalizedDateIfISODateTime(endDate, lng)
+
+  return pipe(
+    localizedStartDate,
+    E.fold(
+      () => handleLocalizedEndDate(localizedEndDate),
+      localizedStartDate =>
+        pipe(
+          localizedEndDate,
+          E.fold(
+            () => <FormattedDateSpan label="from" localizedDate={localizedStartDate} />,
+            localizedEndDate =>
+              handleDateRange(startDate, endDate, localizedStartDate, localizedEndDate)
```

Comment:
It's a little bit weird to me to check for the date range validity at the end of the process after converting the ISO dates to locales.

I would have expected to check first if the dates are valid (e.g. are they valid regarding the conversion and the range?) and that to "fail fast" by returning the `InvalidDateFallback` as soon as possible.

This allows in particular to avoid unnecessary transformations and especially to prevent from giving four parameters to the `handleDateRange` function if the range of dates is not valid from the start.

What do you think?

Reply:
Yes it is more simple to read when the date range is validated first. Having as much as parameters in a function is something that I wanted to avoid but didn't think about moving the range validation at the beginning of the sequence. Thanks for the advice, I updated the code accordingly, and created a function to handle both dates.